android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java

```diff
@@ -26,51 +28,145 @@
+     *
+     * @param filename filename for the archive
+     * @param archive  archive to be queued
+     * @return information about file that produced by the archive
+     */
+    public Single<UploadFile> queueUpload(String filename, Archive archive) {
+        return Single.fromCallable(() -> persist(filename, archive));
+    }
 
-        Single<UploadSession> uploadSessionSingle = uploadFileSingle
-                .flatMap(this::requestUploadSession);
+    /**
+     * This triggers an attempt to upload all the previously queued upload files
+     * You must call queueUpload before this method will have anything to operate on
+     *
+     * @return Observable with information on if the upload was successful or not
+     */
+    public Observable<UploadValidationStatus> uploadToS3() {
+        Observable<String> filenameObservable = Observable.from(uploadDAO
+                .listUploadFilenames());
+
+        return filenameObservable
+                .flatMap(filename -> uploadToS3(filename).toObservable());
+    }
+
+    /**
+     * Uploads a previously queued upload.
+     *
+     * @param filename filename to upload
+     * @return Single of current upload validation status, which could be null if pending upload not found
+     */
+    public Single<UploadValidationStatus> uploadToS3(String filename) {
+        UploadFile uploadFile = uploadDAO.getUploadFile(filename);
+        if (uploadFile == null) {
+            LOG.warn("Could not find uploadFile for filename: " + filename);
+
+            return Single.just(null);
+        }
 
-        return Single.zip(uploadFileSingle, uploadSessionSingle, this::uploadToS3).flatMap(i -> i);
+        return getUploadSession(uploadFile)
+                .flatMap(uploadSession -> getUploadValidationStatus(uploadSession.getId()))
+                .flatMap((uploadValidationStatus) -> {
+                    if (uploadValidationStatus == null) {
+                        LOG.info("Could not find uploadValidationStatus for filename: " + filename);
+                        return Single.just(null);
+                    }
+
+                    if (UploadStatus.SUCCEEDED == uploadValidationStatus.getStatus()) {
+                        LOG.info("Cleaning up successful upload with filename: " + filename
+                                + ", id: " + uploadValidationStatus.getId());
+                        uploadDAO.removeUploadAndSession(filename);
+                        return Single.just(null);
+                    } else if (UploadStatus.VALIDATION_FAILED == uploadValidationStatus.getStatus()) {
+                        LOG.info("Cleaning up validation-failed upload with filename: " + filename
+                                + ", id: " + uploadValidationStatus.getId());
+                        uploadDAO.removeUploadAndSession(filename);
+                        return Single.just(null);
+                    }
+
+                    return getUploadSession(uploadFile)
```

Comment:
This return is crazy complex.  Can you add comments on why you would do the entire upload flow again here?